File: backend/routes/broker_stalker.py
"""Broker Stalker API routes."""
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel
from typing import Optional, List
from modules.broker_stalker_analyzer import BrokerStalkerAnalyzer
from db import BrokerStalkerRepository

router = APIRouter(prefix="/api/broker-stalker", tags=["Broker Stalker"])

analyzer = BrokerStalkerAnalyzer()
repo = BrokerStalkerRepository()


class AddBrokerRequest(BaseModel):
    broker_code: str
    broker_name: Optional[str] = None
    description: Optional[str] = None


class SyncRequest(BaseModel):
    ticker: Optional[str] = None
    days: int = 7


@router.get("/watchlist")
async def get_watchlist():
    """
    Get all brokers in the watchlist.
    
    Returns:
        List of brokers with their power levels
    """
    try:
        watchlist = repo.get_watchlist()
        return {
            "status": "success",
            "count": len(watchlist),
            "brokers": watchlist
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/watchlist")
async def add_broker_to_watchlist(request: AddBrokerRequest):
    """
    Add a broker to the watchlist.
    
    Args:
        request: Broker details
    
    Returns:
        Success status
    """
    try:
        success = repo.add_broker_to_watchlist(
            request.broker_code,
            request.broker_name,
            request.description
        )
        
        if success:
            power_level = analyzer.calculate_power_level(request.broker_code)
            
            return {
                "status": "success",
                "message": f"Broker {request.broker_code} added to watchlist",
                "broker_code": request.broker_code,
                "power_level": power_level
            }
        else:
            raise HTTPException(status_code=400, detail="Failed to add broker")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/watchlist/{broker_code}")
async def remove_broker_from_watchlist(broker_code: str):
    """
    Remove a broker from the watchlist.
    
    Args:
        broker_code: Broker code to remove
    
    Returns:
        Success status
    """
    try:
        success = repo.remove_broker_from_watchlist(broker_code)
        
        if success:
            return {
                "status": "success",
                "message": f"Broker {broker_code} removed from watchlist"
            }
        else:
            raise HTTPException(status_code=404, detail="Broker not found")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/portfolio/{broker_code}")
async def get_broker_portfolio(
    broker_code: str,
    min_net_value: float = Query(default=0, description="Minimum net value filter")
):
    """
    Get broker's current portfolio (active positions).
    
    Args:
        broker_code: Broker code
        min_net_value: Minimum net value to include
    
    Returns:
        Portfolio positions
    """
    try:
        portfolio = repo.get_broker_portfolio(broker_code, min_net_value)
        
        total_net = sum(p['total_net_value'] for p in portfolio)
        
        return {
            "status": "success",
            "broker_code": broker_code,
            "total_positions": len(portfolio),
            "total_net_value": total_net,
            "portfolio": portfolio
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/analysis/{broker_code}/{ticker}")
async def get_broker_analysis(
    broker_code: str,
    ticker: str,
    lookback_days: int = Query(default=30, ge=1, le=90, description="Days to analyze")
):
    """
    Get detailed analysis of broker activity on a ticker.
    
    Args:
        broker_code: Broker code
        ticker: Stock symbol
        lookback_days: Number of days to analyze
    
    Returns:
        Detailed analysis with volumes, streak, and status
    """
    try:
        analysis = analyzer.analyze_broker_activity(broker_code, ticker, lookback_days)
        
        if "error" in analysis:
            raise HTTPException(status_code=500, detail=analysis["error"])
        
        return {
            "status": "success",
            "analysis": analysis
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/chart/{broker_code}/{ticker}")
async def get_broker_chart_data(
    broker_code: str,
    ticker: str,
    days: int = Query(default=7, ge=1, le=30, description="Days of chart data")
):
    """
    Get chart data for broker activity visualization.
    
    Args:
        broker_code: Broker code
        ticker: Stock symbol
        days: Number of days
    
    Returns:
        Time series data for charting
    """
    try:
        chart_data = analyzer.get_daily_chart_data(broker_code, ticker, days)
        
        return {
            "status": "success",
            "broker_code": broker_code,
            "ticker": ticker,
            "days": days,
            "data": chart_data
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/ledger/{broker_code}/{ticker}")
async def get_execution_ledger(
    broker_code: str,
    ticker: str,
    limit: int = Query(default=10, ge=1, le=50, description="Number of records")
):
    """
    Get execution history ledger.
    
    Args:
        broker_code: Broker code
        ticker: Stock symbol
        limit: Maximum records
    
    Returns:
        Recent execution history
    """
    try:
        ledger = analyzer.get_execution_ledger(broker_code, ticker, limit)
        
        return {
            "status": "success",
            "broker_code": broker_code,
            "ticker": ticker,
            "ledger": ledger
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/sync/{broker_code}")
async def sync_broker_data(broker_code: str, request: SyncRequest):
    """
    Sync broker data from done_detail records to tracking table.
    
    Args:
        broker_code: Broker code to sync
        request: Sync parameters
    
    Returns:
        Sync result summary
    """
    try:
        result = analyzer.sync_broker_data(
            broker_code,
            request.ticker,
            request.days
        )
        
        return {
            "status": "success",
            "sync_result": result
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/ticker/{ticker}/activity")
async def get_ticker_broker_activity(
    ticker: str,
    days: int = Query(default=7, ge=1, le=30, description="Days to look back")
):
    """
    Get all broker activity for a specific ticker.
    
    Args:
        ticker: Stock symbol
        days: Number of days
    
    Returns:
        All broker activities on the ticker
    """
    try:
        activity = repo.get_ticker_broker_activity(ticker, days)
        
        return {
            "status": "success",
            "ticker": ticker,
            "days": days,
            "activity_count": len(activity),
            "activity": activity
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/power-level/{broker_code}")
async def calculate_power_level(
    broker_code: str,
    lookback_days: int = Query(default=30, ge=7, le=90, description="Days to analyze")
):
    """
    Calculate and update broker power level.
    
    Args:
        broker_code: Broker code
        lookback_days: Days to analyze
    
    Returns:
        Power level score
    """
    try:
        power_level = analyzer.calculate_power_level(broker_code, lookback_days)
        
        return {
            "status": "success",
            "broker_code": broker_code,
            "power_level": power_level,
            "lookback_days": lookback_days
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/services/api/brokerStalker.ts
/**
 * Broker Stalker API client
 */
import { API_BASE_URL, handleResponse } from './base';

export interface BrokerWatchlistItem {
    broker_code: string;
    broker_name?: string | null;
    description?: string | null;
    power_level: number;
    created_at: string;
    updated_at: string;
}

export interface BrokerTrackingRecord {
    broker_code: string;
    ticker: string;
    trade_date: string;
    total_buy: number;
    total_sell: number;
    net_value: number;
    avg_price?: number | null;
    streak_days: number;
    status?: string | null;
    calculated_at: string;
}

export interface BrokerPortfolioPosition {
    ticker: string;
    total_net_value: number;
    trading_days: number;
    last_trade_date: string;
    avg_execution_price?: number | null;
    streak_days: number;
}

export interface BrokerAnalysis {
    broker_code: string;
    ticker: string;
    lookback_days: number;
    total_buy: number;
    total_sell: number;
    net_value: number;
    streak_days: number;
    status: string;
    daily_activity: Array<{
        date: string;
        buy_volume: number;
        sell_volume: number;
        net_value: number;
    }>;
}

export interface ChartDataPoint {
    date: string;
    buy: number;
    sell: number;
    net: number;
}

export interface ExecutionLedgerEntry {
    date: string;
    action: string;
    volume: number;
    avg_price?: number | null;
    status?: string | null;
}

export interface TickerBrokerActivity {
    broker_code: string;
    trade_date: string;
    total_buy: number;
    total_sell: number;
    net_value: number;
    avg_price?: number | null;
    streak_days: number;
    status?: string | null;
}

export const brokerStalkerApi = {
    getWatchlist: async (): Promise<{ status: string; count: number; brokers: BrokerWatchlistItem[] }> => {
        const response = await fetch(`${API_BASE_URL}/api/broker-stalker/watchlist`);
        return await handleResponse(response, { status: 'error', count: 0, brokers: [] });
    },

    addBrokerToWatchlist: async (payload: {
        broker_code: string;
        broker_name?: string;
        description?: string;
    }): Promise<{ status: string; message: string; broker_code: string; power_level: number }> => {
        const response = await fetch(`${API_BASE_URL}/api/broker-stalker/watchlist`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });
        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.detail || 'Failed to add broker to watchlist');
        }
        return await response.json();
    },

    removeBrokerFromWatchlist: async (brokerCode: string): Promise<{ status: string; message: string }> => {
        const response = await fetch(`${API_BASE_URL}/api/broker-stalker/watchlist/${encodeURIComponent(brokerCode)}`, {
            method: 'DELETE'
        });
        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.detail || 'Failed to remove broker from watchlist');
        }
        return await response.json();
    },

    getBrokerPortfolio: async (
        brokerCode: string,
        minNetValue: number = 0
    ): Promise<{
        status: string;
        broker_code: string;
        total_positions: number;
        total_net_value: number;
        portfolio: BrokerPortfolioPosition[];
    }> => {
        const params = new URLSearchParams({ min_net_value: minNetValue.toString() });
        const response = await fetch(
            `${API_BASE_URL}/api/broker-stalker/portfolio/${encodeURIComponent(brokerCode)}?${params}`
        );
        return await handleResponse(response, {
            status: 'error',
            broker_code: brokerCode,
            total_positions: 0,
            total_net_value: 0,
            portfolio: []
        });
    },

    getBrokerAnalysis: async (
        brokerCode: string,
        ticker: string,
        lookbackDays: number = 30
    ): Promise<{ status: string; analysis: BrokerAnalysis }> => {
        const params = new URLSearchParams({ lookback_days: lookbackDays.toString() });
        const response = await fetch(
            `${API_BASE_URL}/api/broker-stalker/analysis/${encodeURIComponent(brokerCode)}/${encodeURIComponent(ticker)}?${params}`
        );
        return await handleResponse(response);
    },

    getChartData: async (
        brokerCode: string,
        ticker: string,
        days: number = 7
    ): Promise<{
        status: string;
        broker_code: string;
        ticker: string;
        days: number;
        data: ChartDataPoint[];
    }> => {
        const params = new URLSearchParams({ days: days.toString() });
        const response = await fetch(
            `${API_BASE_URL}/api/broker-stalker/chart/${encodeURIComponent(brokerCode)}/${encodeURIComponent(ticker)}?${params}`
        );
        return await handleResponse(response, {
            status: 'error',
            broker_code: brokerCode,
            ticker: ticker,
            days: days,
            data: []
        });
    },

    getExecutionLedger: async (
        brokerCode: string,
        ticker: string,
        limit: number = 10
    ): Promise<{
        status: string;
        broker_code: string;
        ticker: string;
        ledger: ExecutionLedgerEntry[];
    }> => {
        const params = new URLSearchParams({ limit: limit.toString() });
        const response = await fetch(
            `${API_BASE_URL}/api/broker-stalker/ledger/${encodeURIComponent(brokerCode)}/${encodeURIComponent(ticker)}?${params}`
        );
        return await handleResponse(response, {
            status: 'error',
            broker_code: brokerCode,
            ticker: ticker,
            ledger: []
        });
    },

    syncBrokerData: async (
        brokerCode: string,
        ticker?: string,
        days: number = 7
    ): Promise<{ status: string; sync_result: unknown }> => {
        const response = await fetch(
            `${API_BASE_URL}/api/broker-stalker/sync/${encodeURIComponent(brokerCode)}`,
            {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ ticker, days })
            }
        );
        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.detail || 'Failed to sync broker data');
        }
        return await response.json();
    },

    getTickerBrokerActivity: async (
        ticker: string,
        days: number = 7
    ): Promise<{
        status: string;
        ticker: string;
        days: number;
        activity_count: number;
        activity: TickerBrokerActivity[];
    }> => {
        const params = new URLSearchParams({ days: days.toString() });
        const response = await fetch(
            `${API_BASE_URL}/api/broker-stalker/ticker/${encodeURIComponent(ticker)}/activity?${params}`
        );
        return await handleResponse(response, {
            status: 'error',
            ticker: ticker,
            days: days,
            activity_count: 0,
            activity: []
        });
    },

    calculatePowerLevel: async (
        brokerCode: string,
        lookbackDays: number = 30
    ): Promise<{
        status: string;
        broker_code: string;
        power_level: number;
        lookback_days: number;
    }> => {
        const params = new URLSearchParams({ lookback_days: lookbackDays.toString() });
        const response = await fetch(
            `${API_BASE_URL}/api/broker-stalker/power-level/${encodeURIComponent(brokerCode)}?${params}`
        );
        return await handleResponse(response);
    }
};
